`Data.Parity` as per issue #1851

jamesmckinna opened the pull request:
UPDATED:
- [x] `Parity` admits a commutative semiring structure (with `0` = `0ℙ`; `1` = `1ℙ` etc.)
- [x] a function `parity` from `Nat` to `Parity` in `Data.Nat.Base`
- [x] a function `toSign` from `Parity` to `Sign`
- [x] a function `fromSign` from `Sign` to `Parity`
- [x] proofs that `(toSign, fromSign)` form an abelian group iso from `(Parity, 0, _+_)` to `(Sign, +, _*_)` 
- [x] miscellaneous tidyings up of `Data.Sign` etc. to reflect new names for algebraic structures/bundles, cf. issue #1844 and in line with review comments on current PR
- [x] `CHANGELOG`

Not TODO (for subsequent PR):
- [x] proof that `parity` is a commutative semiring homomorphism from `Nat` to `Parity` 
- [x] rationalise the many instances in the library where the `0, 1, 2+ n` pattern analysis of a `Nat` argument takes place (esp. wrt `N{-}ary` operations and `Reflection`, as well as the obvious `Binary` and `Nat` examples); but see [this branch](https://github.com/jamesmckinna/agda-stdlib/tree/suc-suc) and PR #1870

MatthewDaggitt commented:
Looks good to me. Left some comments.

To be honest, this looks like a good size PR. If we resolve my comments above, then I'm happy to merge this in and then leave the morphisms to a second PR.

> It raised a question in my mind as well regarding where homomorphism-like properties belong in the library: under the source type, in *.Properties, or somewhere else? I currently have Parity import Sign in order to define the operations, but it strikes me that there's an asymmetry to that decision, which might better be reflected in having subdirectories of the Algebra.* hierarchy which record concrete instances of (Semi)RingHomomorphisms etc. suggestions welcome!

Having `Data.Parity` import `Data.Sign` is absolutely fine. That's how the rest of the library operates. Yes, in theory you could pull everything out to some neutral ground in `Algebra` but it would be very hard to find, and you'd probably end up re-exporting it publicly into the `Data` modules again which would greatly complicate the dependencies.

jamesmckinna commented:
OK, @MatthewDaggitt thanks for the comments, and the suggestion to slice the PR up. 
As for things being hard-to-find... I had hoped there might be instant acclaim/agreement/assent to the suggestion of `Algebra.Homomorphism.AbelianGroup.*` as a place for such things to go, with individual modules concerning one and only one such structure (so that eventually, we might have long list of such things, but none of them 'hard-to-find'?).
As for the need for subsequent re-importing, you may well have a point, but I'd want to see it before commenting further.

MatthewDaggitt commented:
> As for things being hard-to-find... I had hoped there might be instant acclaim/agreement/assent to the suggestion of Algebra.Homomorphism.AbelianGroup.* as a place for such things to go, with individual modules concerning one and only one such structure (so that eventually, we might have long list of such things, but none of them 'hard-to-find'?).

Feel free to open an issue dedicated especially to discussing this issue, with your proposal and we can have a discussion :smile: I agree, it's definitely something worth thinking about!

jamesmckinna commented on src/Data/Parity/Base.agda:
```diff
@@ -0,0 +1,57 @@
+-- Definition
+
+data Parity : Set where
+  eve : Parity
```
Good question. 
You'll be surprised to learn I did agonise over this. 
Pro my choice: two equal length names; no need to rename to avoid clash with existing uses of `even` in `Data.Nat.*` (but what about `odd`, then?)
Contra: evidently idiotic/quixotic
Either way, I also felt the need *immediately* to introduce the pattern synonyms for the sake of the algebraic structure.

MatthewDaggitt replied:
Yup I also considered suggesting that. I would be happy with that :+1:

Taneb replied:
I'm a -1 on `0ℙ` and `1ℙ`. I don't find them insightful names at all.

jamesmckinna replied:
@Taneb Hmmm: `0 is even parity` `1 is odd parity`?

jamesmckinna replied:
I have gone with `0ℙ` and `1ℙ`.

Taneb commented on src/Data/Parity/Base.agda:
```diff
@@ -0,0 +1,57 @@
+
+-- The opposite parity.
+
+opposite : Parity → Parity
```
I don't think so. The negative of an even number is still an even number, so it might be confusing.

jamesmckinna replied:
`opposite` taken directly from `Sign`, so I'd make the same renaming there, too, but for the parsing clash between `-_` and `-` as constructor... grrr :-(. 
So, to follow your lead, I'd be tempted also to (want to) rename the constructors in `Sign` to `1𝕊` `-1𝕊`. 

Similarly, I'd take my proofs of the `-involutive` and `-injective` properties of `opposite` in `Parity.Properties` and replay them in `Sign.Properties` (or even in `Algebra.Properties`, for a redefinition/additional definition of the 'relational' form of being `Involutive`...)

jamesmckinna replied:
I have gone with `_ᵒ` for `opposite`... but I suppose that `_⁻¹` would have done just as well (but is more of a pain to type)

jamesmckinna commented on src/Data/Parity/Instances.agda:
```diff
@@ -0,0 +1,16 @@
+  using (isDecEquivalence)
+
+instance
+  Parity-≡-isDecEquivalence = isDecEquivalence _≟_
```
Again, slavish copying/emulaiton of existing structure in `Sign`. 
What would be the suggested/approved 'own name' here be? `isDecEquivalence-≟`?

jamesmckinna replied:
Resolved now, I think (?)

jamesmckinna commented on src/Data/Parity/Properties.agda:
```diff
@@ -0,0 +1,415 @@
+
+open import Algebra.Bundles
+open import Data.Empty
+open import Data.Parity.Base renaming (Parity to ℙ)
```
Hmm. OK (and `style-guide` is on your side... ;-)) but I've always been (mostly) comfortable with the endless to-and-fro regarding `Data.Nat` and the name of *its* type: `Nat` or `ℕ`? I'd cheerfully use `ℙ` (and hence `𝕊`...) throughout, but that one day someone would want Sierpinski space to be called `𝕊` etc.

jamesmckinna commented on src/Data/Parity/Properties.agda:
```diff
@@ -0,0 +1,415 @@
+
+opposite-injective : ∀ {p q} → opposite p ≡ opposite q → p ≡ q
+opposite-injective {p} {q} eq = begin
+  p ≡⟨ opposite-inverts eq ⟩
```
Sure. I have terrible trouble with my editor ('default' `emacs`) and the library when browsing other reasoning proofs, because the alignment seems off, but only sometimes, not consistently.

Taneb replied:
I've had alignment issues in emacs because the default font used by emacs doesn't do the monospacing properly with the blackboard bold characters (e.g. `ℙ`) so if I'm using qualified names things get weird. Using another monospace font fixes them.

jamesmckinna replied:
Better now?

jamesmckinna commented:
Removed duplicate bundles, and renamed according to the scheme in `Data.Nat`, for consistency's sake, but I have got *thoroughly* confused as to whether the names of various structures/bundles do/should or not include their units.

MatthewDaggitt commented:
> Removed duplicate bundles, and renamed according to the scheme in Data.Nat, for consistency's sake, but I have got thoroughly confused as to whether the names of various structures/bundles do/should or not include their units.

Yes, we're not consistent on this at all. As far as I can see the pattern is that if it's only a single operator (e.g. monoid) then we do, but if multiple operators (e.g. semiring) then we don't.... To be honest I'd be in favour of never having any units, as the units are nearly always uniquely determined by the operators.

MatthewDaggitt commented on src/Data/Parity/Base.agda:
```diff
@@ -0,0 +1,106 @@
+
+
+------------------------------------------------------------------------
+-- Homomorphism from Parity to Sign: here, or somewhere else?
```
Can you remove the `here or somewhere else` comment?

MatthewDaggitt commented on src/Data/Parity/Properties.agda:
```diff
@@ -0,0 +1,401 @@
+  }
+
+------------------------------------------------------------------------
+-- relating Parity and Sign -- where should this go?
```
Can you remove these comments please!

MatthewDaggitt commented on src/Data/Parity/Base.agda:
```diff
@@ -0,0 +1,106 @@
+
+-- The opposite parity.
+
+_ᵒ : Parity → Parity
```
I'm really sorry, but I'm not sure about this new `_ᵒ` name... I suggested `-_` because it has a clear well-defined semantics as "negate" and therefore users can work out what "negating" a parity does. However `_ᵒ` lacks any such common semantics. If you're not keen on `-_` then I'd much prefer we go back to `opposite`?

jamesmckinna replied:
Well, the symbol is 'well-defined' in some contexts... and I shared @Taneb 's aversion to prefix unary negation. 

I think if this choice doesn't fly, I think I'd rather/better go for _^-1

MatthewDaggitt reviewed (commented):
Otherwise looks great!

jamesmckinna commented:
I'm inspired to do the homomorphism proofs, now I have seen an exemplar in @Taneb 's proof for `List.length`, so watch this space.

jamesmckinna commented:
I think I'm going to stop there. The sheer drudgery of the raw bundle approach to the group isomorphism made me shudder as to how much work, however routine/boilerplate, is needed to establish that `parity` is a `SemiringHomomorphism`. :-(

jamesmckinna commented:
I didn't know how to manage all the equation/inequational reasoning in `Nat.Properties` *together with* equational reasoning across the `parity` function: the typechecker kept complaining about how to resolve ambiguities in the `≡-Reasoning`... and it is too late at night now to think sensibly about this... so it is in a module on its own (reflecting my earlier sympathies/inclinations/thoughts on the topic)

MatthewDaggitt commented:
@jamesmckinna I thought we had (twice!) agreed that we weren't going to add anything more to this PR? Please could remove the last commit from branch, so that I can merge it in without having to go put the whole thing through any more review cycles. As discussed, we can then open a new PR with the second set of additions.

jamesmckinna commented:
Ooops. How to I reverse out of this? `git revert <commit-hash>`? Only this last commit needs to go, I think.

jamesmckinna commented:
I think, as well, that the subtlety of how 'agreement' is arrived at as to the size of a PR is clearer to you than to me. And of course, enthusiasm to finish a task carried me away somewhat...

MatthewDaggitt commented:
> Ooops. How to I reverse out of this? git revert <commit-hash>? Only this last commit needs to go, I think.

I tend to use `git rebase -i HEAD^X` where `X` is the number of commits I want to go back. You then delete the offending commit from the opened file, and it should then remove it for you.

> I think, as well, that the subtlety of how 'agreement' is arrived at as to the size of a PR is clearer to you than to me. And of course, enthusiasm to finish a task carried me away somewhat...

So a PR should be opened as a request for comment/approval on a "completed" unit of work. Obviously some changes may be made in response to the comments, but the ideally the core contribution of the PR should not change. The problem with continuously adding to an open PR is that it becomes very difficult to the reviewers to re-review things efficiently, as they don't know how it interacts with what was already in the PR, so they often end up having to look through the previously reviewed stuff again. Yes, it's possible to view the subsequent commits individually, but it is still difficult to then mentally map those on to the final code that is going to be committed.

jamesmckinna commented:
Hope that worked, and apologies for (once again) blundering around/treading on toes.

MatthewDaggitt commented:
No worries, all looks good now! Thanks for making the requested changes :+1:
